portality/api/v1/data_objects/application.py

```diff
@@ -0,0 +1,323 @@
+                        "link": {
+                            "fields": {
+                                "type": {"coerce": "unicode"},
+                                "url": {"coerce": "url"},
+                            },
+                            "required" : ["type", "url"]
+                        },
+
+                        "license": {
+                            "fields": {
+                                "title": {"coerce": "unicode"},
+                                "type": {"coerce": "unicode"},
+                                "url": {"coerce": "unicode"},
+                                "version": {"coerce": "unicode"},
+                                "open_access": {"coerce": "bool"},
+                                "BY": {"coerce": "bool"},
+                                "NC": {"coerce": "bool"},
+                                "ND": {"coerce": "bool"},
+                                "SA": {"coerce": "bool"},
+                                "embedded": {"coerce": "bool"},
+                                "embedded_example_url": {"coerce": "url"},
+                            },
+                            "required" : [
+                                "embedded",
+                                "title",
+                                "type",
+                                "open_access"
+                            ]
+                        }
+                    }
+                },
+
+                "suggestion" : {
+                    "fields" : {
+                        "article_metadata" : {"coerce" : "bool"}
+                    },
+                    "objects" : [
+                        "articles_last_year"
+                    ],
+                    "required" : ["article_metadata", "articles_last_year"],
+
+                    "structs" : {
+                        "articles_last_year" : {
+                            "fields" : {
+                                "count" : {"coerce" : "integer"},
+                                "url" : {"coerce" : "url"}
+                            },
+                            "required" : ["count", "url"]
+                        }
+                    }
+                },
+
+                "admin" : {
+                    "lists" : {
+                        "contact" : {"contains" : "object"}
+                    },
+                    "required" : ["contact"],
+
+                    "structs" : {
+                        "contact": {
+                            "fields" : {
+                                "email" : {"coerce" : "unicode"},
+                                "name" : {"coerce" : "unicode"}
+                            },
+                            "required" : ["name", "email"]
+                        }
+                    }
+                }
+            }
+        }
+
+        super(IncomingApplication, self).__init__(raw, struct=struct, construct_silent_prune=True, expose_data=True)
+
+    def custom_validate(self):
```

Comment:
We would ideally refactor form, api and CSV reapplication validation (the common parts at least) into some form of validation library that can be easily used by the api, the forms and the custom ingestion code. Not now, but we should do fairly soon in a project so we don't end up in a bit of a pickle if validation requirements change. Or we could leave the refactor until they do change of course.